sycl/source/detail/plugin.hpp

```diff
@@ -184,11 +184,71 @@ class plugin {
   void *getLibraryHandle() { return MLibraryHandle; }
   int unload() { return RT::unloadPlugin(MLibraryHandle); }
 
+  // return the index of PiPlatforms.
+  // If not found, add it and return its index.
+  int getPlatformId(RT::PiPlatform Platform) {
+    if (PiPlatforms) {
+      auto It = std::find(PiPlatforms->begin(), PiPlatforms->end(), Platform);
+      if (It != PiPlatforms->end())
+        return It - PiPlatforms->begin();
+
+      PiPlatforms->push_back(Platform);
+      LastDeviceIds.push_back(0);
+      return PiPlatforms->size() - 1;
+    }
+
+    return -1;
+  }
+  // Device ids are consecutive across platforms within a plugin.
+  // We need to return the same starting index for the given platform.
+  // So, instead of returing the last device id of the given platform,
+  // return the last device id of the predecessor platform.
+  int getStartingDeviceId(RT::PiPlatform Platform) {
+    if (!PiPlatforms)
+      PiPlatforms = new std::vector<RT::PiPlatform>;
+    int PlatformId = getPlatformId(Platform);
+    if (PlatformId == 0)
+      return 0;
+
+    return LastDeviceIds[PlatformId - 1];
+  }
+  // set the id of the last device for the given platform
+  void setLastDeviceId(RT::PiPlatform Platform, int Id) {
+    int PlatformId = getPlatformId(Platform);
+    LastDeviceIds[PlatformId] = Id;
+  }
+  // reset all last device ids to zeros
+  void resetLastDeviceIds() {
+    std::fill(LastDeviceIds.begin(), LastDeviceIds.end(), 0);
+  }
+
+  // Do not leave PiPlatform pointers so that they will not be
+  // released twice at the shutdown time.
+  void resetPiPlatforms() {
+    delete PiPlatforms;
+    PiPlatforms = nullptr;
+  }
+
+  bool containsPiPlatform(RT::PiPlatform Platform) {
+    if (PiPlatforms) {
+      for (RT::PiPlatform Plt : *PiPlatforms) {
+        if (Platform == Plt)
+          return true;
+      }
+    }
+    return false;
+  }
+
 private:
   RT::PiPlugin MPlugin;
   backend MBackend;
   void *MLibraryHandle; // the handle returned from dlopen
   std::shared_ptr<std::mutex> TracingMutex;
+  // vector of PiPlatforms that belong to this plugin
+  std::vector<RT::PiPlatform> *PiPlatforms = nullptr;
+  // represents the unique ids of the last device of each platform
+  // index of this vector corresponds to the index in PiPlatforms vector.
+  std::vector<int> LastDeviceIds;
```

Comment:
Any reason to not use `std::map<RT::PiPlatform, /*ID*/int>` here?

Reply:
map<> seems a better fit. I will change.

Reply:
On the second thought, map is not appropriate in this case.
I need a platform index in the vector to get the starting device index.
map does not provide such index so I will keep it as a vecotr.